File: Leetcode_Java/validParentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        
        Map<Character, Character> parenthesesMap = Map.of(
            ')', '(', 
            ']', '[', 
            '}', '{'
        );

        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (parenthesesMap.containsKey(c)) {
                
                if (!stack.isEmpty() && stack.peek() == parenthesesMap.get(c)) {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}

File: Leetcode_Python/validParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parenthesesMap = { ')' : '(', ']' : '[', '}' : '{'}
        parenthesesStack = []

        for c in s:
            if c in parenthesesMap:
                if parenthesesStack and parenthesesStack[-1] == parenthesesMap[c]:
                    parenthesesStack.pop()
                else:
                    return False
            else:
                parenthesesStack.append(c)

        return True if not parenthesesStack else False

        